### PR TITLE
Use transcripts in block verify tests

### PR DESCRIPTION
**Pull request by oxarbitrage**
Part of https://github.com/ZcashFoundation/zebra/issues/610

### CODE REVIEW

File: zebra-consensus/src/block/tests.rs
Comment (yaahc) [inline zebra-consensus/src/block/tests.rs]: if this is only going to be used in `zebra-consensus` we might want to move these utils into `zebra-test`.

File: zebra-consensus/src/block/tests.rs
Comment (yaahc) [inline zebra-consensus/src/block/tests.rs]: using bools here makes it really confusing to infer what the test is doing. At a minimum we should probably be using `enum`s here so that the different modes are named at the call-site.

TIOLI: Even better might be a builder pattern type for constructing a test runner with a `verify().await` at the end to eventually call this `verify` fn. The fun part is coming up with the API. I haven't thought about it a ton but it might be fun for you to do so ill just give you a rough sketch and then you can run with it however you like.

What I'm imagining is something like this

```rust
use zebra_test::utils::Verify;
let block = zebra_test::utils::get_a_block();

zebra_test::utils::Verifier::new(state, verifier)
    .verify(block.and_then(Verify::Block).and_then(Verify::State))
    .await
    .expect("Ok")
    .verify(block.and_then(Verify::State))
    .await
    .expect_err("Repeated block should fail");
```

if you're confused about the `and_then` in the `verify` call it's a pattern I also use in color-eyre for constructing custom sections, so you can checkout the source [here](https://docs.rs/color-eyre/0.5.0/src/color_eyre/section/mod.rs.html#82-112) to see how to write code like that.

@hdevalence is also really good at this sort of API design so he might be a good person to talk to about ideas for this.

**Review by yaahc — CHANGES_REQUESTED** [review]
Looks good so far, I would like to see some changes to the util function to make it clear what the test plan is but that's my only complaint

File: zebra-consensus/src/block/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: I think that could be a good idea however we are going to need a lot of dependencies into the `zebra-test` including but not limited to `zebra-chain`, `zebra-state`, `tower`, `spandoc`, etc.

Comment (yaahc) [reply zebra-consensus/src/block/tests.rs]: I think that will be fine. The `zebra-state` dependency might be an issue but I think as long as we don't use any of the stuff that depends on `zebra-state` in `zebra-state` itself we won't run into the multiple dependency version issue we ran into with `zebra-test` and `zebra-chain`.

We might even want to setup a `zebra_test::prelude` module to make it easier to bring in commonly used items for tests.

File: zebra-consensus/src/block/tests.rs
Comment (hdevalence) [reply zebra-consensus/src/block/tests.rs]: Looking at this code it seems like what it's trying to do is make a transcript of the expected request / responses for the services it's testing, then check that either the state service or the block verifier are behaving correctly.

Instead of writing a specialized function to do this, it might be better to try to use the Transcript from the testing utilities: https://doc.zebra.zfnd.org/zebra_test/transcript/struct.Transcript.html

This isn't well-documented yet (sorry!) but the idea is that, as a Transcript user, you create a list of the expected requests / responses, and the Transcript can then check a given Service against the list you created.   There's an example of this in zebra-state/tests/basic.rs, which uses a `Lazy<Vec<(Request, Response)>>` to build up the transcript list.  Doing it this way would also eliminate the need for the `get_*_block` methods above.

One gap in the Transcript API is that the transcript only contains successful Responses, so it can't yet be used for checking that some request causes an error. This could probably be fixed but I'm not sure what the right way to check the errors is. (I will make a separate issue for that).

Comment (hdevalence) [reply zebra-consensus/src/block/tests.rs]: https://github.com/ZcashFoundation/zebra/issues/676

File: zebra-consensus/src/block/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: Moved in https://github.com/ZcashFoundation/zebra/pull/675/commits/0b8468b12ef61fa6f78a2d85f5ae39e6fa49fd07 and https://github.com/ZcashFoundation/zebra/pull/675/commits/a2103aecb689be278f47d781aaeed86bb81b27af

File: zebra-consensus/src/block/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: The method from @yaahc seems doable to me, not sure how much time it will take. In the other hand the @hdevalence seems to be even better but harder(for me to understand it) so i will need probably even more time on it. It also seems that the transcript api will need to be fixed first. I of course will like to do it that way if this is the optimal. In the meantime i replaced the bool with enum at https://github.com/ZcashFoundation/zebra/pull/675/commits/e48aad55383a73ead2ec0cc6021a0c1ba741358c

Comment (yaahc) [reply zebra-consensus/src/block/tests.rs]: I've mostly finished the process of fixing the transcript in #678. My hope is we can get that merged tonight or tomorrow then you can go ahead and follow henry's suggestion for basing the tests on `Transcripts`, which is definitely what we should prefer.

Comment (hdevalence) [reply zebra-consensus/src/block/tests.rs]: I don't think that changing the transcript API has to be a prerequisite to this, the current API works for checking that certain requests give certain responses, it just can't handle checking that things expected to error really do error.

Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: Ok good, i will start working on this. Thank you both.

Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: Actually i think https://github.com/ZcashFoundation/zebra/pull/678 is needed.

**Comment by teor2345** [comment]
Note: this PR will conflict with the partial restart PR #736

**Comment by dconnolly** [comment]
Depends on:

- [x] [Error handling in Transcripts PR](https://github.com/ZcashFoundation/zebra/pull/678)
- [x] [CheckpointVerifier partial restarts PR](https://github.com/ZcashFoundation/zebra/pull/736)
  - [ ] rebase on the new tests

**Comment by yaahc** [comment]
looks like one of the blockers has been resolved, @oxarbitrage you should be good to rebase this now I think, or it might be better to wait for the error handling in transcripts change to merge, up to you.

**Comment by yaahc** [comment]
@oxarbitrage I'm sorry it took so long but we've now merged https://github.com/ZcashFoundation/zebra/pull/678 so you should be good to go with a rebase and rewrite as a Transcript. Let me know if you have any issues using the error handling API for the Transcript.

**Comment by oxarbitrage** [comment]
That is perfect @yaahc . Working on writing this tests as transcripts now and will probably force push a new branch here.

**Comment by yaahc** [comment]
@oxarbitrage just wanted to ping you on this one to see how progress is going and see if there's anything I can do to help unblock you on this PR.

**Comment by oxarbitrage** [comment]
Sorry, i left it a bit behind to work in some others. I will resume the work here tomorrow and let you know. Thanks.

File: zebra-consensus/src/block/tests.rs
Comment (yaahc) [inline zebra-consensus/src/block/tests.rs]:
```suggestion
            .ok_or_else(|| eyre!("overflow when calculating 3 hours in the future"))
```

Comment (yaahc) [reply zebra-consensus/src/block/tests.rs]: Alternatively its probably safe to use `expect` here because we never expect 3 hours from the current time to cause overflow.

File: zebra-consensus/src/block/tests.rs
Comment (yaahc) [inline zebra-consensus/src/block/tests.rs]: Does this work instead?
```suggestion
        let trans = zebra_test::vectors::DUMMY_TX1.zcash_deserialize_into().unwrap();
        transactions.push(trans);
```

**Review by yaahc — DISMISSED** [review]
LGTM, I just have some minor lints

File: zebra-consensus/src/block/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: Yes, something like this worked, added at https://github.com/ZcashFoundation/zebra/pull/675/commits/2ced9f7cf713b5c83637ece0874cf2630ebcc673

File: zebra-consensus/src/block/tests.rs
Comment (oxarbitrage) [reply zebra-consensus/src/block/tests.rs]: Added at https://github.com/ZcashFoundation/zebra/pull/675/commits/2ced9f7cf713b5c83637ece0874cf2630ebcc673

**Review by yaahc — APPROVED** [review]
awesome, tyvm

**Comment by teor2345** [comment]
Thanks, this is really great work!

Let's think about how to do this for the rest of the verify tests (checkpoint and chain).

**Pull request merged**
